packages/core/strapi/lib/commands/develop.js

```diff
@@ -7,95 +7,113 @@ const chokidar = require('chokidar');
 const execa = require('execa');
 const { getOr } = require('lodash/fp');
 
-const { createLogger } = require('@strapi/logger');
 const loadConfiguration = require('../core/app-configuration');
 const strapi = require('../index');
-const buildAdmin = require('./build');
+const tsUtils = require('../utils/typescript');
+const { buildTypeScript, buildAdmin } = require('./builders');
 
 /**
  * `$ strapi develop`
  *
  */
 module.exports = async function({ build, watchAdmin, polling, browser }) {
-  const dir = process.cwd();
-  const config = loadConfiguration(dir);
-  const logger = createLogger(config.logger, {});
+  let dir = process.cwd();
+
+  if (tsUtils.isTypeScriptProject(dir)) {
+    dir = path.join(dir, 'dist');
+  }
 
   try {
     if (cluster.isMaster || cluster.isPrimary) {
-      const serveAdminPanel = getOr(true, 'admin.serveAdminPanel')(config);
-
-      const buildExists = fs.existsSync(path.join(dir, 'build'));
-      // Don't run the build process if the admin is in watch mode
-      if (build && !watchAdmin && serveAdminPanel && !buildExists) {
-        try {
-          await buildAdmin({ optimization: false, forceBuild: false });
-        } catch (err) {
-          process.exit(1);
-        }
-      }
-
-      if (watchAdmin) {
-        try {
-          execa('npm', ['run', '-s', 'strapi', 'watch-admin', '--', '--browser', browser], {
-            stdio: 'inherit',
-          });
-        } catch (err) {
-          process.exit(1);
-        }
-      }
-
-      cluster.on('message', (worker, message) => {
-        switch (message) {
-          case 'reload':
-            logger.info('The server is restarting\n');
-            worker.send('kill');
-            break;
-          case 'killed':
-            cluster.fork();
-            break;
-          case 'stop':
-            process.exit(1);
-          default:
-            return;
-        }
-      });
-
-      cluster.fork();
+      return masterProcess({ dir, build, browser });
```

Comment:
Could we maybe go `primaryProcess`? Language wise it's more inclusive, which is also the reason why node introduced `cluser.isPrimary`: https://nodejs.org/api/cluster.html#clusterismaster

https://inclusivenaming.org/word-lists/tier-1/#master-slave